### ISSUE TITLE
An async middleware seems to sometimes return an empty response

### ISSUE THREAD

**Issue by rally25rs** [koajs/koa#928]
* Koa 2.0.1
* Node 7.7.1
* OSX

This is a tough one to reproduce as it seems to be some sort of timing issue, but I have an async middleware that in a `while` loop uses `http.request` to poll another server. On a certain response, it resolves the middleware back to Koa. It basically looks like this:

server.js (sets up koa and koa-router)
```
  // router is koa-router
  router.post('/api/submit', require('./routes/submit'));

  koaApplication
    .use(router.routes());
```

routes/submit.js (my middleware that does some polling)
```
const STATUS_QUERY_DELAY = 1000;

async function delay(duration = STATUS_QUERY_DELAY) {
  return new Promise((resolve) => {
    setTimeout(() => resolve(), duration);
  });
}

module.exports = async function(ctx, next) {
  var status = 'Pending';
  var attempt = 0;
  var response;

  while(status === 'Pending' && attempt < STATUS_QUERY_ATTEMPTS) {
    await delay();

    // the "request()" function is just a wrapper around http.request() but wraps
    // it in a Promise so that I can "await" it.
    response = await request({ ... });

    status = // get status from response body
    attempt++;
  }

  ctx.set(response.headers);
  ctx.status = response.status;
  ctx.body = response.body;

  console.log(`Responding with ${ctx.status} - ${ctx.body}`);
  await next();
};
```

The final console.log _always_ prints the right information that I expect, however from a client (both curl and Postman) I _sometimes_ get an empty response.

> curl: (52) Empty reply from server

Changing `STATUS_QUERY_DELAY` seems to affect how often this happens, with `1000` being the worst offender (tried 500 and 1500 and the empty responses happen much less frequently, maybe 1 in 20 attempts).

On attempts where the `while` loop makes >1 iteration, it seems more likely to happen, but not always, and curl does wait multiple seconds, then node.js and curl both output at the same time, node with the "Respondng with" console statement, and curl with "Empty reply"

Is there something I am doing wrong here?

**Comment by jonathanong** [comment]
How long are these requests taking? I'm wondering if you're hitting the `http` server's timeout. also, if you're using `request`, it would be `ctx.status = response.statusCode`.

there's no need for `await next()` at the end of your functions because you've already handled the request.

**Comment by rally25rs** [comment]
The request usually take 1 or 2 loops of the specified delay length, so usually something like 1 to 2 seconds, but even if I kick the delay up to 2000ms and the response takes 2 loops, so 4 seconds, curl still reports success about 90% of the time.

The `response` object there that is returned from `request()` is of my own creation. `request()` returns:

```
        resolve({
          status: res.statusCode,
          body: rawData, // built up from the 'data' events
          headers: res.headers
        });
```

**Comment by rally25rs** [comment]
I noticed the problem gets worse if I try multiple requests in parallel. Then it failed 50% of the time.

> for i in {1..10}; do curl --data "data=something" http://localhost:3006/api/submit & done

Unfortunately the code is proprietary so I can't post the complete project, but I'll see if I can rig up a minimal sample project that still demonstrates the issue...

**Comment by fl0w** [comment]
This seems really weird, and the only thing I can think of is what I don't see - does your promise wrapper around `request` properly reject/throw? Interesting bug. Feels like a timing bug but I can't fathom what causes it.

**Comment by rally25rs** [comment]
I think I have a distilled [sample project that demonstrates the issue](https://github.com/rally25rs/koa-issue-928).

I originally just tried to point it at the GitHub public API, but couldn't get it to fail. I had to make a 2nd server that sometimes doesn't respond 200 Success to get it to fail.

I'm still not sure if this is just an oddity of my code or some deeper issue. Any help is much appreciated!

**Comment by rally25rs** [comment]
OK I did a bunch more debugging on this, going into Koa itself. It seems that in the occurrences where curl/postman gets an empty response, the underlying socket reports that it is not writable.

It gets into Koa's `response.js` `get writable()` and in that getter, `this.res.socket.writable` returns `false`, so no body content is written to it.

**Comment by PlasmaPower** [comment]
> On both the "bad" and "good" sequences, the stack trace from the 'close' event seems to be the same:

Yeah that's how Node events work unfortunately. Here's what I'd recommend instead:

```js
const oldClose = socket.close;
socket.close = function () {
  console.trace('Close called');
  oldClose.apply(socket, arguments);
};
```

If close isn't called, then the socket is probably closing itself. Maybe look into the timeout.

**Comment by rally25rs** [comment]
Well, after a whole lot of time spent on this, I realized that it is indeed a timeout issue.
I have the sockets set to timeout after 2 seconds, and my polling to occur every 0.5 seconds.

My sample/test server was returning `requestNumber++ % 2 ? 'Success' : 'Pending'` so in my head I was like "ok so every second request will end the polling loop, so 2 loops of 0.5 seconds will make my requests take ~1sec each, well under the 2 seconds"

Turns out when you run 10 of them in parallel, those loops don't exactly line up (duh), and they can make anywhere from 1 to 4 polling iterations to complete, which could push it over the 2 second mark.

Very strangely, it was dead consistent that 1 in 10 would fail, every time, across dozens of runs.

So, all this time it was just a dumb timeout :)
